hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
-        // TODO: delete those completed/failed rows
+        match self.cleanup_impl().await {
+            Ok(_) => {}
+            Err(error) => {
+                error!(error = ?error, "WebhookCleaner::cleanup failed");
+            }
+        }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::config;
+    use crate::kafka_producer::{create_kafka_producer, KafkaContext};
+    use rdkafka::mocking::MockCluster;
+    use rdkafka::producer::{DefaultProducerContext, FutureProducer};
+    use sqlx::PgPool;
+
+    const APP_METRICS_TOPIC: &str = "app_metrics";
+
+    async fn create_mock_kafka() -> (
+        MockCluster<'static, DefaultProducerContext>,
+        FutureProducer<KafkaContext>,
+    ) {
+        let cluster = MockCluster::new(1).expect("failed to create mock brokers");
+
+        let config = config::KafkaConfig {
+            kafka_producer_linger_ms: 0,
+            kafka_producer_queue_mib: 50,
+            kafka_message_timeout_ms: 5000,
+            kafka_compression_codec: "none".to_string(),
+            kafka_hosts: cluster.bootstrap_servers(),
+            app_metrics_topic: APP_METRICS_TOPIC.to_string(),
+            plugin_log_entries_topic: "plugin_log_entries".to_string(),
+            kafka_tls: false,
+        };
+
+        (
+            cluster,
+            create_kafka_producer(&config)
+                .await
+                .expect("failed to create mocked kafka producer"),
+        )
+    }
+
+    #[sqlx::test(migrations = "../migrations", fixtures("webhook_cleanup"))]
+    async fn test_cleanup_impl(db: PgPool) {
+        let (mock_cluster, mock_producer) = create_mock_kafka().await;
+        mock_cluster
+            .create_topic(APP_METRICS_TOPIC, 1, 1)
+            .expect("failed to create mock app_metrics topic");
+
+        let table_name = "job_queue";
+        let queue_name = "webhooks";
+
+        let webhook_cleaner = WebhookCleaner::new_from_pool(
+            &queue_name,
+            &table_name,
+            db,
+            mock_producer,
+            APP_METRICS_TOPIC.to_owned(),
+        )
+        .expect("unable to create webhook cleaner");
+
+        let _ = webhook_cleaner
+            .cleanup_impl()
+            .await
+            .expect("webbook cleanup_impl failed");
+
+        // TODO: I spent a lot of time trying to get the mock Kafka consumer to work, but I think
+        // I've identified an issue with the rust-rdkafka library:
+        //   https://github.com/fede1024/rust-rdkafka/issues/629#issuecomment-1863555417
+        //
+        // I wanted to test the messages put on the AppMetrics topic, but I think we need to figure
+        // out that issue about in order to do so. (Capture uses the MockProducer but not a
+        // Consumer, fwiw.)
+        //
+        // For now, I'll probably have to make `cleanup_impl` return the row information so at
+        // least we can inspect that for correctness.
```

Comment:
I don't know if it would be more work, but we can always spin-up a kafka in the docker compose stack, so no mocking is required.

Reply:
That's true, I may check that out today.

Reply:
Figured out the mocking, needed `.set("auto.offset.reset", "earliest")`